### PR TITLE
Convert exception on sync client to java exception

### PR DESCRIPTION
**Pull request by beeender**
- Update realm-sync to 1.10.1
- update object-store to 1e3cbb1789
- Convert exception on sync client to java exception
  This will give us some better information when excpetion happens
  on the sync client thread.

### CODE REVIEW

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 1]
```diff
@@ -44,7 +50,23 @@ struct AndroidClientListener : public realm::BindingCallbackThreadObserver {
...
         JniUtils::detach_current_thread();
     }
-} s_client_thread_listener;
+
+    void handle_error(std::exception const& e) override
+    {
+        JNIEnv* env = JniUtils::get_env(true);
+        std::string msg = format("An exception has been thrown on the sync client thread:\n%1", e.what());
+        Log::f(msg.c_str());
+        // Since user has no way to handle exceptions thrown on the sync client thread, we just convert it to an Java
```
Comment (nhachicha) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: to a Java

**Comment by nhachicha** [comment]
you have a compile error
```
  /tmp/realm-java/realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp:54:10: error: 'void AndroidClientListener::handle_error(const std::exception&)' marked override, but does not override
       void handle_error(std::exception const& e) override
```
did you update the object store commit to point to realm/realm-object-store#456 ?

**Comment by beeender** [comment]
@nhachicha No i didn't ... unfortunately object-store master needs `1.9.1` whose corresponding ROS release is not in public yet. We might need to branch out the Object Store for a while to merge this ....

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 2]
```diff
@@ -25,11 +25,17 @@
 #include "util.hpp"
 #include "jni_util/jni_utils.hpp"
 #include "jni_util/java_method.hpp"
+#include "jni_util/java_class.hpp"
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: Aren't these supposed to be in alpha order?

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 3]
```diff
@@ -25,11 +25,17 @@
...
 
 using namespace realm;
+using namespace realm::util;
 using namespace realm::jni_util;
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: alpha order?

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 4]
```diff
@@ -44,7 +50,23 @@ struct AndroidClientListener : public realm::BindingCallbackThreadObserver {
...
+        std::string msg = format("An exception has been thrown on the sync client thread:\n%1", e.what());
+        Log::f(msg.c_str());
+        // Since user has no way to handle exceptions thrown on the sync client thread, we just convert it to an Java
+        // exception to get more debug information for ourself.
+        // FIXME: We really need to find a universal and clever way to get the native backtrace when exception thrown
+        env->ThrowNew(m_realm_exception_class, msg.c_str());
+    }
+
+private:
+    // For some reasons, FindClass() doesn't work in the native thread even when the jvm is attached before. Get the
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: jvm or JVM.  pick one.

**Comment by beeender** [comment]
@realm/java This did a core update as well.

**Comment by beeender** [comment]
This goes to master since it won't be compatible with ROS < 1.6.0 which is considered as a breaking change i think?

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 5]
```diff
@@ -44,7 +50,23 @@ struct AndroidClientListener : public realm::BindingCallbackThreadObserver {
...
+        std::string msg = format("An exception has been thrown on the sync client thread:\n%1", e.what());
+        Log::f(msg.c_str());
+        // Since user has no way to handle exceptions thrown on the sync client thread, we just convert it to an Java
+        // exception to get more debug information for ourself.
+        // FIXME: We really need to find a universal and clever way to get the native backtrace when exception thrown
+        env->ThrowNew(m_realm_exception_class, msg.c_str());
+    }
+
+private:
+    // For some reasons, FindClass() doesn't work in the native thread even when the JVM is attached before. Get the
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: Is it a bug in Android? Is it a know bug? Or does it also apply to any JVM implementation?

Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: no idea, I cannot find it documented anywhere. I feel like it is an Android only thing. Strange thing is, if you have the method id and the jclass/jobject, the java method can be called without any problems.

**Comment by kneth** [comment]
Should we disable the failing unit test if we cannot get a core hot fix?

Otherwise I think we can merge it.

**Comment by beeender** [comment]
@kneth I guess we cannot ... It will make the master not-releasable ...

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,8 @@
 
 ### Breaking Changes
 
+* [ObjectServer] Update protocol version to 18 which is only compatible with ROS > 1.6.0.
```
Comment (dalinaum) [inline CHANGELOG.md]: Updated protocol?

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 7]
```diff
@@ -72,8 +94,9 @@ JNIEXPORT void JNICALL Java_io_realm_SyncManager_nativeInitializeSyncManager(JNI
         JStringAccessor base_file_path(env, sync_base_dir); // throws
         SyncManager::shared().configure_file_system(base_file_path, SyncManager::MetadataMode::NoEncryption);
 
+        static AndroidClientListener client_thread_listner(env);
```
Comment (dalinaum) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: It should be `client_thread_listener`.

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp [hunk 8]
```diff
@@ -72,8 +94,9 @@ JNIEXPORT void JNICALL Java_io_realm_SyncManager_nativeInitializeSyncManager(JNI
...
         // Register Sync Client thread start/stop callback
-        g_binding_callback_thread_observer = &s_client_thread_listener;
+        g_binding_callback_thread_observer = &client_thread_listner;
```
Comment (dalinaum) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: `client_thread_listner`

File: realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp (hunk 5 above)
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp]: It is documented here: https://developer.android.com/training/articles/perf-jni.html#faq_FindClass

**Pull request merged**